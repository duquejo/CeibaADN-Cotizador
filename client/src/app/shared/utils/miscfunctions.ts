import moment from 'moment';
import { ISelectionOptions } from '../../feature/Admin/models/FormSelector';

export const obtenerCalendarioActivo = (
    calendarioActivo: string | number | null = null, 
    todosCalendarios: ISelectionOptions[] ) => {

    const estaActivoEnCalendarios = todosCalendarios.some( 
        ( calendario: ISelectionOptions ) => {
        return Number( calendario.value ) === Number( calendarioActivo );
    });

    if( calendarioActivo === '' || ! estaActivoEnCalendarios ) {
      if( todosCalendarios.length > 0 && todosCalendarios[0].value ) {
        return Number( todosCalendarios[0].value );
      } else {
        return null;
      }
    }
    return Number( calendarioActivo );
};

export const extraerIdsDeArray = ( needle: Array<any>, haystack: Array<any> ) => {
  
  if( ! needle.length ) {
    return [];
  }

  const needleIds = needle.map( ( elemento: any ) => {
    if( ! elemento.hasOwnProperty( 'id' ) ){
      return elemento;
    }
    return elemento.id;
  } );

  return haystack.filter( ( elemento: any ) => {
    return ( needleIds.indexOf( elemento!.id ) !== -1 );
  });
};

export const cotizacionTemplate = ( cotizacion: any, allInfo = false ) => {

  const identificador = allInfo ? '<th>Identificador</th>' : '';
  const centroVacacionalId = allInfo ? `<td>${ cotizacion.centroVacacional?.id }</td>` : '';
  const centroVacacionalCodigo = allInfo ? `<td>${ cotizacion.centroVacacional?.codigo }</td>` : '';
  const categoriaUsuariosId = allInfo ? `<td>${ cotizacion.categoriaUsuarios.id }</td>` : '';
  const categoriaUsuariosDescripcion = cotizacion.categoriaUsuarios?.descripcion ? cotizacion.categoriaUsuarios.descripcion : '';
  const centroVacacionalDescripcion = cotizacion.centroVacacional?.descripcion ? cotizacion.centroVacacional.descripcion : '';

  return `<table class="resultados__table">
            <caption>Información general</caption>
            <thead>
              <tr>
                <th>Identificador</th>
                <th>Código</th>
                <th>Fecha de inicio</th>
                <th>Fecha de finalización</th>
                <th># de personas</th>
                <th>Total grupo</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>${ cotizacion.id }</td>
                <td>${ cotizacion.codigo }</td>
                <td>${ moment( cotizacion.fechaInicio ).format('YYYY-MM-DD') }</td>
                <td>${ moment( cotizacion.fechaFin ).format('YYYY-MM-DD') }</td>
                <td>${ cotizacion.personas }</td>
                <td>$ ${ cotizacion.total }</td>
              </tr>
            </tbody>
          </table>
          <table class="resultados__table">
            <caption>Información del centro vacacional</caption>
            <thead>
              <tr>
                ${ identificador }
                <th>Nombre</th>
                <th>Descripción</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                ${ centroVacacionalId }
                ${ centroVacacionalCodigo }
                <td>${ cotizacion.centroVacacional?.nombre }</td>
                <td>${ centroVacacionalDescripcion }</td>
              </tr>
            </tbody>
            <table class="resultados__table">
              <caption>Información de la categoría de usuario</caption>
              <thead>
                <tr>
                  ${ identificador }
                  <th>Nombre</th>
                  <th>Descripción</th>
                  <th>Valor de día temporada alta</th>
                  <th>Valor de día temporada baja</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  ${ categoriaUsuariosId }
                  <td>${ cotizacion.categoriaUsuarios.nombre }</td>
                  <td>${ categoriaUsuariosDescripcion }</td>
                  <td>$ ${ cotizacion.categoriaUsuarios.valorAlta }</td>
                  <td>$ ${ cotizacion.categoriaUsuarios.valorBaja }</td>
                </tr>
              </tbody>                
            </table>
          </table>`;
};
