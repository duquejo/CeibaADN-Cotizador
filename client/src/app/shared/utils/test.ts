export const setTextEvent = (identifier: string, text: string) => ({
  persist: () => {},
  target: {
    name: identifier,
    value: text,
  },
});
