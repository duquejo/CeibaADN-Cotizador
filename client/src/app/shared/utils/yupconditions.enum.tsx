export enum yupConditions {
    minStringLength=2,
    maxStringLength=50
}