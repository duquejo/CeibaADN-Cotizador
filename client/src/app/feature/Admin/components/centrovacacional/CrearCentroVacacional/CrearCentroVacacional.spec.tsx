import * as React from 'react';
import { RenderResult, fireEvent, render, wait } from '@testing-library/react';
import { SinonStub, stub } from 'sinon';
import { CrearCentroVacacional } from '.';
import { setTextEvent } from 'app/shared/utils/test';

describe('CrearCentroVacacional test', () => {
    
  let componentWrapper: RenderResult;
  let componentProps: React.ComponentProps<typeof CrearCentroVacacional> & {
    onSubmit: SinonStub;
  };

  beforeEach(() => {
    componentProps = {
      onSubmit: stub(),
      optionCalendarios: [
        {
          id: 1,
          nombre: 'Calendario de prueba festivos',
          descripcion: 'Descripción de prueba',
          festivos: [ 
            '2021-12-18T00:00:00-05:00',
            '2022-02-22T00:00:00-05:00'
          ]
        },
        {
          id: 3,
          nombre: 'Calendario de prueba festivos II',
          festivos: [ 
            '2022-02-11T00:00:00-05:00',
            '2022-06-18T00:00:00-05:00',
            '2022-04-03T00:00:00-05:00',
          ]
        }
      ],
      optionCategoriasUsuarios: [
        {
          id: 5,
          nombre: 'Categoría de usuarios de prueba',
          descripcion: 'Descripción de categoría de usuarios',
          valorAlta: 64000,
          valorBaja: 30000
        },
        {
          id: 12,
          nombre: 'Otra categoría de prueba',
          valorAlta: 32000,
          valorBaja: 12000
        }
      ]
    };
    componentWrapper = render(<CrearCentroVacacional { ...componentProps } />);
  });

  it('Debería coincidir con el snapshot', () => {
    expect( componentWrapper.container ).toMatchSnapshot();
  });

  it('Debería fallar al no tener los valores obligatorios presentes', async () => {

    // Arrange
    const elem = componentWrapper.container;
    const submitButton = elem.querySelector('button[type="submit"]');

    // Act
    await wait(() => {
      submitButton && fireEvent.click( submitButton );
    });

    // Assert
    const spans = elem.querySelectorAll('span');
    expect( spans.length ).toBe(3);
    expect( spans[0].textContent ).toBe('El campo título es requerido');
  });

  it('Debería realizar el submit con la información básica', async () => {

    // Arrange
    const elem = componentWrapper.container;
    const title = elem.querySelector('input[name="title"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    // Act
    await wait(() => {
      title && fireEvent.change(title, setTextEvent('title', 'Centro vacacional de prueba'));
    });     
    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    // Assert
    const formSubmitted = componentProps.onSubmit.firstCall.args[0];
    expect( formSubmitted.nombre ).toBe('Centro vacacional de prueba');
    expect( formSubmitted.descripcion ).toBe('');
    expect( formSubmitted.calendarioActivo ).toBe(null);
    expect( typeof formSubmitted.calendarios ).toBe('object');
    expect( typeof formSubmitted.categoriaUsuarios ).toBe('object');
  });

  it('Debería realizar el submit con la descripción y los datos obligatorios', async () => {

    // Arrange
    const elem = componentWrapper.container;
    const title = elem.querySelector('input[name="title"]');
    const description = elem.querySelector('textarea[name="description"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    // Act
    await wait(() => {
      title && fireEvent.change(title, setTextEvent('title', 'Centro vacacional de prueba'));
    });
    await wait(() => {
      description && fireEvent.change(description, setTextEvent('description', 'Descripción del calendario de prueba'));
    });
    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    // Assert
    const formSubmitted = componentProps.onSubmit.firstCall.args[0];
    expect( formSubmitted.nombre ).toBe('Centro vacacional de prueba');
    expect( formSubmitted.descripcion ).toBe('Descripción del calendario de prueba');
    expect( formSubmitted.calendarioActivo ).toBe(null);
    expect( typeof formSubmitted.calendarios ).toBe('object');
    expect( typeof formSubmitted.categoriaUsuarios ).toBe('object');
  });
});