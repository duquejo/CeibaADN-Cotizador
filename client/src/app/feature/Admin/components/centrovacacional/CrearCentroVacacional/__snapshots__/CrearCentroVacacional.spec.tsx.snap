// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrearCentroVacacional test Debería coincidir con el snapshot 1`] = `
<div>
  <form
    class="cv__form"
  >
    <h3>
      Crear centro vacacional
    </h3>
    <label
      for="cv-title"
    >
      <b>
        Título 
        <div
          class="required"
        >
          *
        </div>
      </b>
    </label>
    <input
      class="sc-gsDKAQ kdvfBj"
      id="cv-title"
      name="title"
      placeholder="Nombre del centro vacacional, ejemplo: 'Cartagena Resort'."
      value=""
    />
    <label
      for="cv-description"
    >
      <b>
        Descripción
      </b>
    </label>
    <textarea
      class="sc-jRQBWg hlzhTV"
      id="cv-description"
      name="description"
      placeholder="Descripción del centro vacacional, ejemplo: 'Pasajes paradisiacos en la costa Colombiana'."
      rows="3"
    />
    <b>
      Asignación de calendarios de temporadas altas
    </b>
    <small
      class="sc-hKwDye hDvPSz"
    >
      Dicha asignación le permitirá a los administradores tener una visión de los calendarios presentes para sus centros vacacionales y configuración de días de manera dinámica.
    </small>
    <div
      class="rmsc selector__container calendarios__container"
    >
      <div
        aria-labelledby="Seleccionar"
        aria-readonly="true"
        class="dropdown-container"
        tabindex="0"
      >
        <div
          class="dropdown-heading"
        >
          <div
            class="dropdown-heading-value"
          >
            <span
              class="gray"
            >
              Selecciona un calendario de festivos o días de alta
            </span>
          </div>
          <svg
            class="dropdown-heading-dropdown-arrow gray"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-width="2"
            width="24"
          >
            <path
              d="M6 9L12 15 18 9"
            />
          </svg>
        </div>
      </div>
    </div>
    <b>
      Asignación de categorías de usuarios
    </b>
    <small
      class="sc-hKwDye hDvPSz"
    >
      Dicha asignación le permitirá a los clientes obtener opciones personalizadas de precios según su nivel salarial.
    </small>
    <div
      class="rmsc selector__container categorias__container"
    >
      <div
        aria-labelledby="Seleccionar"
        aria-readonly="true"
        class="dropdown-container"
        tabindex="0"
      >
        <div
          class="dropdown-heading"
        >
          <div
            class="dropdown-heading-value"
          >
            <span
              class="gray"
            >
              Selecciona las categorías de usuarios
            </span>
          </div>
          <svg
            class="dropdown-heading-dropdown-arrow gray"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-width="2"
            width="24"
          >
            <path
              d="M6 9L12 15 18 9"
            />
          </svg>
        </div>
      </div>
    </div>
    <label
      for="calendarioActivo"
    >
      <b>
        Asignación de calendario activo
      </b>
    </label>
    <small
      class="sc-hKwDye hDvPSz"
    >
      El calendario activo será el que el cotizador tendrá visible y presente en los cálculos.
    </small>
    <select
      class="sc-dkPtRN gkhtzY"
      id="active_calendar"
      name="calendarioActivo"
    >
      <option
        value=""
      >
        Selecciona un calendario del listado
      </option>
    </select>
    <button
      class="sc-bdvvtL ibsxaf"
      type="submit"
    >
      Crear
    </button>
  </form>
</div>
`;
